File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.cpp
class Solution {
private:
    void countPossible(string & tiles, string curr, unordered_map<int,int> &flag , unordered_set<string> &possible){
        for(int i=0;i<tiles.size();i++){
            if(flag[i]!=1){
                curr.push_back(tiles[i]);
                flag[i]=1;
                possible.insert(curr);
                countPossible(tiles,curr,flag,possible);
                curr.pop_back();
                flag[i]=0;
            }
        }
    }
public:
    int numTilePossibilities(string tiles) {
        unordered_set<string> possible;
        unordered_map<int,int> flag;
        countPossible(tiles,"",flag,possible);
        return (int) possible.size();
    }
};

File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.rs
use std::collections::HashSet;
use std::collections::HashMap;
impl Solution {
    fn count_possible(tiles :&String, curr : String,
        flag :&mut HashMap<usize,i32>, possible : &mut HashSet<String>){
            
            for(i,c) in tiles.chars().enumerate(){
                if flag.get(&i) != Some(&1){
                    let mut new_curr = curr.clone();
                    new_curr.push(c);
                    flag.insert(i,1);
                    possible.insert(new_curr.clone());
                    Solution::count_possible(tiles,new_curr,flag,possible);
                    flag.insert(i,0);                    
                }
            }
    }
    pub fn num_tile_possibilities(tiles: String) -> i32 {
        let mut possible : HashSet<String> = HashSet::new();
        let mut flag : HashMap<usize,i32> = HashMap::new();
        Solution::count_possible(&tiles,"".to_string(), &mut flag, &mut possible);
        possible.len() as i32
    }
}